File: Fundamentals_level_5/Localmind/frontend/src/lib/api/studio/blogs.ts
/**
 * Blog Posts API
 * Educational Note: Handles AI-generated comprehensive blog posts with images.
 * Blog posts are SEO-optimized and stored as markdown files with images.
 */

import axios from 'axios';
import { API_BASE_URL } from '../client';
import type { JobStatus } from './index';

/**
 * Blog types available for generation
 */
export type BlogType =
  | 'case_study'
  | 'listicle'
  | 'how_to_guide'
  | 'opinion'
  | 'product_review'
  | 'news'
  | 'tutorial'
  | 'comparison'
  | 'interview'
  | 'roundup';

/**
 * Blog outline section from planning phase
 */
export interface BlogOutlineSection {
  heading: string;
  content_description: string;
  subsections?: string[];
  needs_image: boolean;
  image_description?: string;
}

/**
 * Generated blog image info
 */
export interface BlogImage {
  purpose: string;
  section_heading: string;
  filename: string;
  placeholder: string;
  alt_text: string;
  url: string;
}

/**
 * Blog post generation job
 */
export interface BlogJob {
  id: string;
  source_id: string;
  source_name: string;
  direction: string;
  target_keyword: string;
  blog_type: BlogType;
  status: JobStatus;
  status_message: string;
  error_message: string | null;
  // Blog plan fields
  title: string | null;
  meta_description: string | null;
  outline: BlogOutlineSection[];
  target_word_count: number;
  tone: string | null;
  // Generated content
  images: BlogImage[];
  markdown_file: string | null;
  markdown_url: string | null;
  preview_url: string | null;
  word_count: number | null;
  // Metadata
  iterations: number | null;
  input_tokens: number | null;
  output_tokens: number | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

/**
 * Response from starting blog post generation
 */
export interface StartBlogResponse {
  success: boolean;
  job_id?: string;
  status?: string;
  message?: string;
  error?: string;
}

/**
 * Response from getting blog job status
 */
export interface BlogJobStatusResponse {
  success: boolean;
  job?: BlogJob;
  error?: string;
}

/**
 * Response from listing blog jobs
 */
export interface ListBlogJobsResponse {
  success: boolean;
  jobs: BlogJob[];
  error?: string;
}

/**
 * Blog Posts API
 */
export const blogsAPI = {
  /**
   * Start blog post generation via blog agent
   */
  async startGeneration(
    projectId: string,
    sourceId: string,
    direction?: string,
    targetKeyword?: string,
    blogType?: BlogType
  ): Promise<StartBlogResponse> {
    try {
      const response = await axios.post(
        `${API_BASE_URL}/projects/${projectId}/studio/blog`,
        {
          source_id: sourceId,
          direction: direction || '',
          target_keyword: targetKeyword || '',
          blog_type: blogType || 'how_to_guide',
        }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      console.error('Error starting blog post generation:', error);
      throw error;
    }
  },

  /**
   * Get the status of a blog post job
   */
  async getJobStatus(projectId: string, jobId: string): Promise<BlogJobStatusResponse> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/blog-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      console.error('Error getting blog job status:', error);
      throw error;
    }
  },

  /**
   * List all blog post jobs for a project
   */
  async listJobs(projectId: string, sourceId?: string): Promise<ListBlogJobsResponse> {
    try {
      const params = sourceId ? { source_id: sourceId } : {};
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/blog-jobs`,
        { params }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      console.error('Error listing blog jobs:', error);
      throw error;
    }
  },

  /**
   * Delete a blog post job
   */
  async deleteJob(projectId: string, jobId: string): Promise<{ success: boolean; error?: string }> {
    try {
      const response = await axios.delete(
        `${API_BASE_URL}/projects/${projectId}/studio/blog-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      console.error('Error deleting blog job:', error);
      throw error;
    }
  },

  /**
   * Get the full URL for a blog file (markdown or image)
   */
  getFileUrl(projectId: string, filename: string): string {
    return `${API_BASE_URL}/projects/${projectId}/studio/blogs/${filename}`;
  },

  /**
   * Get the preview URL for a blog post (returns markdown content)
   */
  getPreviewUrl(projectId: string, jobId: string): string {
    return `${API_BASE_URL}/projects/${projectId}/studio/blogs/${jobId}/preview`;
  },

  /**
   * Get the download URL for a blog post (ZIP with markdown + images)
   */
  getDownloadUrl(projectId: string, jobId: string): string {
    return `${API_BASE_URL}/projects/${projectId}/studio/blogs/${jobId}/download`;
  },

  /**
   * Fetch markdown preview content directly
   */
  async getPreview(projectId: string, jobId: string): Promise<string> {
    try {
      const response = await axios.get(this.getPreviewUrl(projectId, jobId), {
        responseType: 'text',
      });
      return response.data;
    } catch (error) {
      console.error('Error fetching blog preview:', error);
      throw error;
    }
  },

  /**
   * Poll blog job status until complete or error
   * Educational Note: Added initial delay to avoid race condition where
   * the job might not be saved to disk yet when polling starts.
   */
  async pollJobStatus(
    projectId: string,
    jobId: string,
    onProgress?: (job: BlogJob) => void,
    intervalMs: number = 2000,
    maxAttempts: number = 150 // Blog posts can take longer with images
  ): Promise<BlogJob> {
    let attempts = 0;
    let currentInterval = intervalMs;

    // Initial delay to let the job be created and saved
    await new Promise((resolve) => setTimeout(resolve, 1000));

    while (attempts < maxAttempts) {
      const response = await this.getJobStatus(projectId, jobId);

      // Handle race condition: job might not exist yet in first few attempts
      if (!response.success || !response.job) {
        if (attempts < 3) {
          // Give the backend more time to create the job
          await new Promise((resolve) => setTimeout(resolve, currentInterval));
          attempts++;
          continue;
        }
        throw new Error(response.error || 'Failed to get job status');
      }

      const job = response.job;

      if (onProgress) {
        onProgress(job);
      }

      if (job.status === 'ready' || job.status === 'error') {
        return job;
      }

      await new Promise((resolve) => setTimeout(resolve, currentInterval));

      attempts++;

      // Gradually increase interval for long-running jobs
      if (attempts > 5 && currentInterval < 5000) {
        currentInterval = Math.min(currentInterval * 1.2, 5000);
      }
    }

    throw new Error('Blog post generation timed out');
  },
};

File: Fundamentals_level_5/Localmind/backend/app/api/studio/blogs.py
"""
Blog Post endpoints - AI-generated comprehensive blog posts.

Educational Note: Blog posts demonstrate agent-based generation with image support:
1. blog_agent_executor orchestrates the generation
2. Claude creates markdown structure and content
3. Gemini generates images for hero and sections
4. Complete package: Markdown + images

Agent Pattern:
- Uses blog_agent_executor for orchestration
- Agent has tools for planning, image generation, and writing
- Multi-step process with intermediate results
- Final output is a complete markdown blog post

Output Structure:
- Markdown file with frontmatter (title, meta_description, etc.)
- Image files for hero and section illustrations
- All files stored in project's studio/blogs folder
- ZIP download available for full package

Routes:
- POST /projects/<id>/studio/blog                        - Start generation
- GET  /projects/<id>/studio/blog-jobs/<id>              - Job status
- GET  /projects/<id>/studio/blog-jobs                   - List jobs
- GET  /projects/<id>/studio/blogs/<file>                - Serve file
- GET  /projects/<id>/studio/blogs/<id>/preview          - Preview markdown
- GET  /projects/<id>/studio/blogs/<id>/download         - Download ZIP
- DELETE /projects/<id>/studio/blog-jobs/<id>            - Delete job
"""
import io
import zipfile
from pathlib import Path
from flask import jsonify, request, current_app, send_file, Response
from app.api.studio import studio_bp
from app.services.studio_services import studio_index_service
from app.services.tool_executors.blog_agent_executor import blog_agent_executor
from app.utils.path_utils import get_studio_dir


@studio_bp.route('/projects/<project_id>/studio/blog', methods=['POST'])
def generate_blog_post(project_id: str):
    """
    Start blog post generation via blog agent.

    Request Body:
        - source_id: UUID of the source to generate blog from (required)
        - direction: User's direction/guidance (optional)
        - target_keyword: SEO keyword/phrase to target (optional)
        - blog_type: Category of blog post (optional, default: how_to_guide)

    Response:
        - 202 Accepted with job_id for polling
    """
    try:
        data = request.get_json()

        # Validate input
        source_id = data.get('source_id')
        if not source_id:
            return jsonify({
                'success': False,
                'error': 'source_id is required'
            }), 400

        direction = data.get('direction', '')
        target_keyword = data.get('target_keyword', '')
        blog_type = data.get('blog_type', 'how_to_guide')

        # Validate blog_type
        valid_blog_types = [
            'case_study', 'listicle', 'how_to_guide', 'opinion',
            'product_review', 'news', 'tutorial', 'comparison',
            'interview', 'roundup'
        ]
        if blog_type not in valid_blog_types:
            blog_type = 'how_to_guide'

        # Execute via blog_agent_executor (creates job and launches agent)
        result = blog_agent_executor.execute(
            project_id=project_id,
            source_id=source_id,
            direction=direction,
            target_keyword=target_keyword,
            blog_type=blog_type
        )

        if not result.get('success'):
            return jsonify(result), 400

        return jsonify({
            'success': True,
            'job_id': result['job_id'],
            'status': result['status'],
            'message': result['message']
        }), 202

    except Exception as e:
        current_app.logger.error(f"Error starting blog post generation: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to start blog post generation: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/blog-jobs/<job_id>', methods=['GET'])
def get_blog_job_status(project_id: str, job_id: str):
    """
    Get the status of a blog post generation job.

    Response:
        - Job object with status, progress, and generated content
    """
    try:
        job = studio_index_service.get_blog_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Blog job {job_id} not found'
            }), 404

        return jsonify({
            'success': True,
            'job': job
        })

    except Exception as e:
        current_app.logger.error(f"Error getting blog job status: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@studio_bp.route('/projects/<project_id>/studio/blog-jobs', methods=['GET'])
def list_blog_jobs(project_id: str):
    """
    List all blog post jobs for a project.

    Query Parameters:
        - source_id: Optional filter by source

    Response:
        - List of blog jobs (newest first)
    """
    try:
        source_id = request.args.get('source_id')
        jobs = studio_index_service.list_blog_jobs(project_id, source_id)

        return jsonify({
            'success': True,
            'jobs': jobs
        })

    except Exception as e:
        current_app.logger.error(f"Error listing blog jobs: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500


@studio_bp.route('/projects/<project_id>/studio/blogs/<filename>', methods=['GET'])
def get_blog_file(project_id: str, filename: str):
    """
    Serve a blog file (markdown or image).

    Response:
        - Markdown file or image file with appropriate headers
    """
    try:
        blog_dir = get_studio_dir(project_id) / "blogs"
        filepath = blog_dir / filename

        if not filepath.exists():
            return jsonify({
                'success': False,
                'error': f'File not found: {filename}'
            }), 404

        # Validate the file is within the expected directory (security)
        try:
            filepath.resolve().relative_to(blog_dir.resolve())
        except ValueError:
            return jsonify({
                'success': False,
                'error': 'Invalid file path'
            }), 400

        # Determine mimetype
        if filename.endswith('.md'):
            mimetype = 'text/markdown'
        elif filename.endswith('.png'):
            mimetype = 'image/png'
        elif filename.endswith('.jpg') or filename.endswith('.jpeg'):
            mimetype = 'image/jpeg'
        else:
            mimetype = 'application/octet-stream'

        return send_file(
            filepath,
            mimetype=mimetype,
            as_attachment=False
        )

    except Exception as e:
        current_app.logger.error(f"Error serving blog file: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to serve file: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/blogs/<job_id>/preview', methods=['GET'])
def preview_blog_post(project_id: str, job_id: str):
    """
    Serve blog post markdown for preview.

    Response:
        - Markdown file content
    """
    try:
        # Get job to find markdown file
        job = studio_index_service.get_blog_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Blog job {job_id} not found'
            }), 404

        markdown_file = job.get('markdown_file')
        if not markdown_file:
            return jsonify({
                'success': False,
                'error': 'Blog post not yet generated'
            }), 404

        blog_dir = get_studio_dir(project_id) / "blogs"
        filepath = blog_dir / markdown_file

        if not filepath.exists():
            return jsonify({
                'success': False,
                'error': f'Markdown file not found: {markdown_file}'
            }), 404

        # Read and return markdown content
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()

        return Response(
            content,
            mimetype='text/markdown',
            headers={'Content-Type': 'text/markdown; charset=utf-8'}
        )

    except Exception as e:
        current_app.logger.error(f"Error previewing blog post: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to preview blog post: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/blogs/<job_id>/download', methods=['GET'])
def download_blog_post(project_id: str, job_id: str):
    """
    Download blog post as ZIP file (markdown + images).

    Response:
        - ZIP file containing markdown and all images
    """
    try:
        # Get job to find files
        job = studio_index_service.get_blog_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Blog job {job_id} not found'
            }), 404

        markdown_file = job.get('markdown_file')
        if not markdown_file:
            return jsonify({
                'success': False,
                'error': 'Blog post not yet generated'
            }), 404

        blog_dir = get_studio_dir(project_id) / "blogs"

        # Create ZIP in memory
        zip_buffer = io.BytesIO()
        with zipfile.ZipFile(zip_buffer, 'w', zipfile.ZIP_DEFLATED) as zip_file:
            # Add markdown file
            markdown_path = blog_dir / markdown_file
            if markdown_path.exists():
                zip_file.write(markdown_path, markdown_file)

            # Add image files
            images = job.get('images', [])
            for image_info in images:
                image_filename = image_info.get('filename')
                if image_filename:
                    image_path = blog_dir / image_filename
                    if image_path.exists():
                        zip_file.write(image_path, f"images/{image_filename}")

        zip_buffer.seek(0)

        # Generate filename from title
        title = job.get('title', 'blog_post')
        safe_name = "".join(c for c in title if c.isalnum() or c in (' ', '_', '-')).strip()[:50]
        zip_filename = f"{safe_name}.zip"

        return send_file(
            zip_buffer,
            mimetype='application/zip',
            as_attachment=True,
            download_name=zip_filename
        )

    except Exception as e:
        current_app.logger.error(f"Error downloading blog post: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to download blog post: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/blog-jobs/<job_id>', methods=['DELETE'])
def delete_blog_job(project_id: str, job_id: str):
    """
    Delete a blog post job and its files.

    Response:
        - Success status
    """
    try:
        # Get job to find files
        job = studio_index_service.get_blog_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Blog job {job_id} not found'
            }), 404

        # Delete files
        blog_dir = get_studio_dir(project_id) / "blogs"

        # Delete markdown file
        markdown_file = job.get('markdown_file')
        if markdown_file:
            markdown_path = blog_dir / markdown_file
            if markdown_path.exists():
                markdown_path.unlink()

        # Delete image files
        images = job.get('images', [])
        for image_info in images:
            image_filename = image_info.get('filename')
            if image_filename:
                image_path = blog_dir / image_filename
                if image_path.exists():
                    image_path.unlink()

        # Delete job from index
        studio_index_service.delete_blog_job(project_id, job_id)

        return jsonify({
            'success': True,
            'message': f'Blog job {job_id} deleted'
        })

    except Exception as e:
        current_app.logger.error(f"Error deleting blog job: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to delete blog job: {str(e)}'
        }), 500
